subsys/net/ip/ipv4_autoconf.c

```diff
@@ -0,0 +1,270 @@
+			break;
+		}
+
+		/* success, add new IPv4 address */
+		if (!net_if_ipv4_addr_add(iface,
+					  &net_if_get_config(iface)->ipv4auto.requested_ip,
+					  NET_ADDR_AUTOCONF, 0)) {
+			NET_DBG("Failed to add IPv4 addr to iface %p", iface);
+			return;
+		}
+
+		net_if_get_config(iface)->ipv4auto.state = NET_IPV4_AUTOCONF_CONFIGURED;
+		break;
+	default:
+		break;
+	}
+}
+
+static inline uint32_t ipv4_autoconf_get_timeout(struct net_if *iface)
+{
+	switch (net_if_get_config(iface)->ipv4auto.state) {
+	case NET_IPV4_AUTOCONF_PROBE:
+		if (net_if_get_config(iface)->ipv4auto.conflict_cnt >=
+		    IPV4_AUTOCONF_MAX_CONFLICTS) {
+			NET_DBG("rate limiting");
+			return K_SECONDS(IPV4_AUTOCONF_RATE_LIMIT_INTERVAL);
+		} else if (net_if_get_config(iface)->ipv4auto.probe_cnt ==
+			   IPV4_AUTOCONF_PROBE_NUM) {
+			return K_SECONDS(IPV4_AUTOCONF_ANNOUNCE_INTERVAL);
+		}
+
+		return IPV4_AUTOCONF_PROBE_WAIT *
+		       CONFIG_SYS_CLOCK_TICKS_PER_SEC +
+		       (sys_rand32_get() % MSEC_PER_SEC);
+
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+		return K_SECONDS(IPV4_AUTOCONF_ANNOUNCE_INTERVAL);
+
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static void send_probe(struct net_if *iface)
+{
+	struct net_pkt *probe = prepare_arp(iface);
+
+	if (!probe) {
+		NET_DBG("Failed to prepare probe %p", iface);
+		return;
+	}
+
+	NET_DBG("Probing");
+	net_if_queue_tx(iface, probe);
+	net_if_get_config(iface)->ipv4auto.probe_cnt++;
+	net_if_get_config(iface)->ipv4auto.state = NET_IPV4_AUTOCONF_PROBE;
+
+	k_delayed_work_init(&net_if_get_config(iface)->ipv4auto.timeout,
+			    ipv4_autoconf_timeout);
+	k_delayed_work_submit(&net_if_get_config(iface)->ipv4auto.timeout,
+			      ipv4_autoconf_get_timeout(iface));
+}
+
+static void send_announcement(struct net_if *iface)
+{
+	struct net_pkt *announce = prepare_arp(iface);
+
+	if (!announce) {
+		NET_DBG("Failed to prepare announcement %p", iface);
+		return;
+	}
+
+	NET_DBG("Announcing");
+	net_if_queue_tx(iface, announce);
+
+	net_if_get_config(iface)->ipv4auto.announce_cnt++;
+	net_if_get_config(iface)->ipv4auto.state = NET_IPV4_AUTOCONF_ANNOUNCE;
+
+	k_delayed_work_submit(&net_if_get_config(iface)->ipv4auto.timeout,
+			      ipv4_autoconf_get_timeout(iface));
+}
+
+void net_ipv4_autoconf_start(struct net_if *iface)
+{
+	/* initialize interface and start probing */
+	net_if_get_config(iface)->ipv4auto.state = NET_IPV4_AUTOCONF_INIT;
+	net_if_get_config(iface)->ipv4auto.probe_cnt = 0;
+	net_if_get_config(iface)->ipv4auto.announce_cnt = 0;
+	net_if_get_config(iface)->ipv4auto.conflict_cnt = 0;
+	memset(&net_if_get_config(iface)->ipv4auto.current_ip, 0, sizeof(struct in_addr));
+	net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[0] = 169;
+	net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[1] = 254;
+	net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[2] =
+		(uint8_t)(sys_rand32_get() % 254);
+	net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[3] =
+		(uint8_t)(sys_rand32_get() % 254);
+
```

Comment:
I suggest here a debug print like this

```
	NET_DBG("Starting probe for 169.254.%d.%d",
		net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[2],
		net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[3]);

```

otherwise it is a bit difficult to see what is being done.